From ddf3da49b3356c0086419a4763e516c12ce95783 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 1 Dec 2025 18:48:28 +0100
Subject: [PATCH 22/49] win32u: Avoid a crash when drawable fails to be
 created.

Signed-off-by: sewn <sewn@disroot.org>
---
 dlls/win32u/opengl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/win32u/opengl.c b/dlls/win32u/opengl.c
index 57a0c77..8fdc7de 100644
--- a/dlls/win32u/opengl.c
+++ b/dlls/win32u/opengl.c
@@ -1656,7 +1656,7 @@ static BOOL context_sync_drawables( struct wgl_context *context, HDC draw_hdc, H
     if (draw_hdc && !context->draw) context->draw = get_dc_opengl_drawable( draw_hdc );
     if (read_hdc && !context->read) context->read = get_dc_opengl_drawable( read_hdc );
 
-    new_draw = get_updated_drawable( draw_hdc, context->format, context->draw );
+    if (!(new_draw = get_updated_drawable( draw_hdc, context->format, context->draw ))) return FALSE;
     if (!draw_hdc && context->draw == context->read) opengl_drawable_add_ref( (new_read = new_draw) );
     else if (draw_hdc && draw_hdc == read_hdc) opengl_drawable_add_ref( (new_read = new_draw) );
     else new_read = get_updated_drawable( read_hdc, context->format, context->read );
-- 
2.52.0

