From fdd54c371149253e67c8a83f14369495ea393e26 Mon Sep 17 00:00:00 2001
From: sewn <sewn@disroot.org>
Date: Fri, 26 Dec 2025 17:11:53 +0300
Subject: [PATCH 26/26] wine.inf: Disable unused services

Signed-off-by: sewn <sewn@disroot.org>
---
 loader/wine.inf.in | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 80a4a062351..da23557b263 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -1099,7 +1099,6 @@ StartType=3
 ErrorControl=1
 
 [Services]
-HKLM,%CurrentVersion%\RunServices,"winemenubuilder",2,"%11%\winemenubuilder.exe -a -r"
 HKLM,"System\CurrentControlSet\Services\Dnscache\Parameters",,16
 HKLM,"System\CurrentControlSet\Services\Eventlog\Application",,16
 HKLM,"System\CurrentControlSet\Services\Eventlog\System","Sources",0x10000,""
@@ -1219,10 +1218,8 @@ services,"@%11%\ws2_32.dll,-4"
 [InfFiles]
 input.inf,"@%12%\hidclass.sys,-1"
 mouhid.inf,"@%12%\mouhid.sys,-1"
-winebth.inf,"@%12%\winebth.sys,-1"
 winebus.inf,"@%12%\winebus.sys,-1"
 winehid.inf,"@%12%\winehid.sys,-1"
-wineusb.inf,"@%12%\wineusb.sys,-1"
 winexinput.inf,"@%12%\winexinput.sys,-1"
 
 [NlsFiles]
-- 
2.52.0

