From 1843de52f50347728e5ed76d99b026cbaf2c108e Mon Sep 17 00:00:00 2001
From: Giraffe1966 <35208168+Giraffe1966@users.noreply.github.com>
Date: Sun, 10 Dec 2023 06:34:51 +0000
Subject: [PATCH 1/4] Make roblox cursor lock properly.

---
 dlls/winewayland.drv/wayland_pointer.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_pointer.c b/dlls/winewayland.drv/wayland_pointer.c
index c20ba170285..fb13399d22c 100644
--- a/dlls/winewayland.drv/wayland_pointer.c
+++ b/dlls/winewayland.drv/wayland_pointer.c
@@ -748,7 +748,7 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
                                               BOOL covers_vscreen)
 {
     struct wayland_pointer *pointer = &process_wayland.pointer;
-    BOOL needs_relative, needs_lock, needs_confine;
+    BOOL needs_relative, needs_lock, needs_confine, is_one_pixel;;
     static unsigned int once;
 
     if (!process_wayland.zwp_pointer_constraints_v1)
@@ -758,9 +758,10 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
         return;
     }
 
+    is_one_pixel = confine_rect && (int)(confine_rect->right - confine_rect->left)==1 && (int)(confine_rect->bottom - confine_rect->top)==1;
     needs_lock = wl_surface && (confine_rect || covers_vscreen) &&
-                 !pointer->cursor.wl_surface;
-    needs_confine = wl_surface && confine_rect && pointer->cursor.wl_surface;
+                 (!pointer->cursor.wl_surface || is_one_pixel);
+    needs_confine = wl_surface && confine_rect && pointer->cursor.wl_surface && !is_one_pixel;
 
     if (!needs_confine && pointer->zwp_confined_pointer_v1)
     {
@@ -842,7 +843,7 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
         return;
     }
 
-    needs_relative = !pointer->cursor.wl_surface &&
+    needs_relative = (!pointer->cursor.wl_surface || is_one_pixel) &&
                      pointer->constraint_hwnd &&
                      pointer->constraint_hwnd == pointer->focused_hwnd;
 
-- 
2.46.2


From 32b0fccb746a675206639b6c990fed9a28829d10 Mon Sep 17 00:00:00 2001
From: Giraffe1966 <35208168+Giraffe1966@users.noreply.github.com>
Date: Sun, 10 Dec 2023 10:27:18 +0000
Subject: [PATCH 2/4] Make relative motion smoother at lower mouse movement
 speeds.

---
 dlls/winewayland.drv/wayland_pointer.c | 6 +++---
 dlls/winewayland.drv/waylanddrv.h      | 2 ++
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_pointer.c b/dlls/winewayland.drv/wayland_pointer.c
index fb13399d22c..aa67a7ba722 100644
--- a/dlls/winewayland.drv/wayland_pointer.c
+++ b/dlls/winewayland.drv/wayland_pointer.c
@@ -261,18 +261,18 @@ static void relative_pointer_v1_relative_motion(void *private,
     HWND hwnd;
     POINT screen;
     struct wayland_win_data *data;
+    struct wayland_pointer *pointer = &process_wayland.pointer;
 
     if (!(hwnd = wayland_pointer_get_focused_hwnd())) return;
     if (!(data = wayland_win_data_get(hwnd))) return;
 
     wayland_surface_coords_to_window(data->wayland_surface,
-                                     wl_fixed_to_double(dx),
-                                     wl_fixed_to_double(dy),
+                                     wl_fixed_to_double(dx)+pointer->pendingX,
+                                     wl_fixed_to_double(dy)+pointer->pendingY,
                                      (int *)&screen.x, (int *)&screen.y);
 
     wayland_win_data_release(data);
 
-
     input.type = INPUT_MOUSE;
     input.mi.dx = screen.x;
     input.mi.dy = screen.y;
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index 3dee8e432df..cd89f4959b1 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -100,6 +100,8 @@ struct wayland_pointer
     HWND constraint_hwnd;
     uint32_t enter_serial;
     uint32_t button_serial;
+    double pendingX;
+    double pendingY;
     struct wayland_cursor cursor;
     pthread_mutex_t mutex;
 };
-- 
2.46.2


From 3074e20ea7df45f4a588697aa1728d3ac3c65e37 Mon Sep 17 00:00:00 2001
From: Giraffe1966 <35208168+Giraffe1966@users.noreply.github.com>
Date: Mon, 11 Dec 2023 20:31:26 +0000
Subject: [PATCH 3/4] Prevent left click from disabling cursor lock.

---
 dlls/win32u/input.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 14862e9a8a4..6f903a84acd 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -2568,7 +2568,7 @@ BOOL clip_fullscreen_window( HWND hwnd, BOOL reset )
     if (!is_window_rect_full_screen( &rect, dpi )) return FALSE;
     if (is_captured_by_system()) return FALSE;
     if (NtGetTickCount() - thread_info->clipping_reset < 1000) return FALSE;
-    if (!reset && clipping_cursor && thread_info->clipping_cursor) return FALSE;  /* already clipping */
+    if (!reset && clipping_cursor) return FALSE;  /* already clipping */
 
     ctx = set_thread_dpi_awareness_context( NTUSER_DPI_PER_MONITOR_AWARE );
     monitor_info = monitor_info_from_window( hwnd, MONITOR_DEFAULTTONEAREST );
-- 
2.46.2


From 8a5ef9a58f5c6a1c82746b6e78169c330b5a2937 Mon Sep 17 00:00:00 2001
From: Giraffe1966 <35208168+Giraffe1966@users.noreply.github.com>
Date: Sun, 17 Dec 2023 06:37:18 +0000
Subject: [PATCH 4/4] Ensure the pointer is warped to the confinement region
 when WAYLAND_ClipCursor is called.

---
 dlls/winewayland.drv/wayland_pointer.c | 132 +++++++++++++++++++------
 dlls/winewayland.drv/waylanddrv.h      |   2 +
 2 files changed, 102 insertions(+), 32 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_pointer.c b/dlls/winewayland.drv/wayland_pointer.c
index aa67a7ba722..26308df2a78 100644
--- a/dlls/winewayland.drv/wayland_pointer.c
+++ b/dlls/winewayland.drv/wayland_pointer.c
@@ -92,11 +92,84 @@ static void pointer_handle_motion_internal(wl_fixed_t sx, wl_fixed_t sy)
     NtUserSendHardwareInput(hwnd, 0, &input, 0);
 }
 
+// Calling thread should hold pointer lock
+static void update_pointer_surface_coords(wl_fixed_t sx, wl_fixed_t sy)
+{
+    struct wayland_pointer *pointer = &process_wayland.pointer;
+
+    pointer->sx = wl_fixed_to_double(sx);
+    pointer->sy = wl_fixed_to_double(sy);
+}
+
+/*
+ Try to warp the pointer to the specified confines.
+ This is necessary since pointer warping won't necessarily happen
+ when making a call to zwp_pointer_constraints_v1_{confine_pointer/lock_pointer} with the region set.
+ In fact, most compositors won't warp the pointer when the region is set on a constraint.
+
+ This method of warping done by this function isn't guaranteed to warp the pointer either (the spec doesn't provide any guarentees that it will,
+ only that warping may occur), but this method seems to work in more compositors than setting the region of a constraint.
+
+ Calling thread should own the pointer mutex. This function warps the pointer
+ to the confinement's nearest edge.
+ There should also be no currently acive cursor constraint--if there is one, it should be destroyed and then recreated aftewards
+*/
+static void warp_to_confines(struct wl_surface *wl_surface, RECT *confine_rect)
+{
+    double sx, sy;
+    struct wayland_pointer *pointer = &process_wayland.pointer;
+
+    sx = pointer->sx;
+    sy = pointer->sy;
+
+    if (!confine_rect || (confine_rect->top<=sy && confine_rect->bottom >= sy && confine_rect->left <= sx && confine_rect->right >= sx))
+    {
+        // cursor is already in the confinement area
+        return;
+    }
+
+    if (confine_rect->top>sy)
+    {
+        sy = confine_rect->top;
+    }
+    if (confine_rect->bottom<sy)
+    {
+        sy = confine_rect->bottom;
+    }
+    if (confine_rect->left>sx)
+    {
+        sx = confine_rect->left;
+    }
+    if (confine_rect->right<sx)
+    {
+        sx = confine_rect->right;
+    }
+
+    struct zwp_locked_pointer_v1 *zwp_locked_pointer_v1 =
+        zwp_pointer_constraints_v1_lock_pointer(
+                    process_wayland.zwp_pointer_constraints_v1,
+                    wl_surface,
+                    pointer->wl_pointer,
+                    NULL,
+                    ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
+
+    zwp_locked_pointer_v1_set_cursor_position_hint(zwp_locked_pointer_v1, wl_fixed_from_double(sx), wl_fixed_from_double(sy));
+    wl_surface_commit(wl_surface);
+
+    zwp_locked_pointer_v1_destroy(zwp_locked_pointer_v1);
+
+    TRACE("Warping cursor to: (%f,%f)\n", sx, sy);
+}
+
 static void pointer_handle_motion(void *data, struct wl_pointer *wl_pointer,
                                   uint32_t time, wl_fixed_t sx, wl_fixed_t sy)
 {
     struct wayland_pointer *pointer = &process_wayland.pointer;
 
+    pthread_mutex_lock(&pointer->mutex);
+    update_pointer_surface_coords(sx, sy);
+    pthread_mutex_unlock(&pointer->mutex);
+
     /* Ignore absolute motion events if in relative mode. */
     if (pointer->zwp_relative_pointer_v1) return;
 
@@ -122,6 +195,8 @@ static void pointer_handle_enter(void *data, struct wl_pointer *wl_pointer,
     pthread_mutex_lock(&pointer->mutex);
     pointer->focused_hwnd = hwnd;
     pointer->enter_serial = serial;
+
+    update_pointer_surface_coords(sx, sy);
     pthread_mutex_unlock(&pointer->mutex);
 
     /* The cursor is undefined at every enter, so we set it again with
@@ -789,24 +864,18 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
                       confine_rect->right - confine_rect->left,
                       confine_rect->bottom - confine_rect->top);
 
-        if (!pointer->zwp_confined_pointer_v1 || pointer->constraint_hwnd != hwnd)
-        {
-            if (pointer->zwp_confined_pointer_v1)
-                zwp_confined_pointer_v1_destroy(pointer->zwp_confined_pointer_v1);
-            pointer->zwp_confined_pointer_v1 =
-                zwp_pointer_constraints_v1_confine_pointer(
-                    process_wayland.zwp_pointer_constraints_v1,
-                    wl_surface,
-                    pointer->wl_pointer,
-                    region,
-                    ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
-            pointer->constraint_hwnd = hwnd;
-        }
-        else
-        {
-            zwp_confined_pointer_v1_set_region(pointer->zwp_confined_pointer_v1,
-                                               region);
-        }
+        if (pointer->zwp_confined_pointer_v1)
+            zwp_confined_pointer_v1_destroy(pointer->zwp_confined_pointer_v1);
+        if (confine_rect)
+            warp_to_confines(wl_surface, confine_rect);
+        pointer->zwp_confined_pointer_v1 =
+            zwp_pointer_constraints_v1_confine_pointer(
+                process_wayland.zwp_pointer_constraints_v1,
+                wl_surface,
+                pointer->wl_pointer,
+                region,
+                ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
+        pointer->constraint_hwnd = hwnd;
 
         TRACE("Confining to hwnd=%p wayland=%d,%d+%d,%d\n",
               pointer->constraint_hwnd,
@@ -820,20 +889,19 @@ static void wayland_pointer_update_constraint(struct wl_surface *wl_surface,
     {
         HWND hwnd = wl_surface_get_user_data(wl_surface);
 
-        if (!pointer->zwp_locked_pointer_v1 || pointer->constraint_hwnd != hwnd)
-        {
-            if (pointer->zwp_locked_pointer_v1)
-                zwp_locked_pointer_v1_destroy(pointer->zwp_locked_pointer_v1);
-            pointer->zwp_locked_pointer_v1 =
-                zwp_pointer_constraints_v1_lock_pointer(
-                    process_wayland.zwp_pointer_constraints_v1,
-                    wl_surface,
-                    pointer->wl_pointer,
-                    NULL,
-                    ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
-            pointer->constraint_hwnd = hwnd;
-            TRACE("Locking to hwnd=%p\n", pointer->constraint_hwnd);
-        }
+        if (pointer->zwp_locked_pointer_v1)
+            zwp_locked_pointer_v1_destroy(pointer->zwp_locked_pointer_v1);
+        if (confine_rect)
+            warp_to_confines(wl_surface, confine_rect);
+        pointer->zwp_locked_pointer_v1 =
+            zwp_pointer_constraints_v1_lock_pointer(
+                process_wayland.zwp_pointer_constraints_v1,
+                wl_surface,
+                pointer->wl_pointer,
+                NULL,
+                ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT);
+        pointer->constraint_hwnd = hwnd;
+        TRACE("Locking to hwnd=%p\n", pointer->constraint_hwnd);
     }
 
     if (!process_wayland.zwp_relative_pointer_manager_v1)
diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
index cd89f4959b1..52b963968a7 100644
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -102,6 +102,8 @@ struct wayland_pointer
     uint32_t button_serial;
     double pendingX;
     double pendingY;
+    double sx;
+    double sy;
     struct wayland_cursor cursor;
     pthread_mutex_t mutex;
 };
-- 
2.46.2

